### PR TITLE
Prevent heap buffer overflow

### PR DESCRIPTION
**Pull request by eustas**

### CODE REVIEW

File: lib/jxl/dec_group.cc [hunk 1]
```diff
@@ -259,6 +259,8 @@ Status DecodeGroupImpl(GetBlock* JXL_RESTRICT get_block,
     r[i] =
         Rect(block_rect.x0() >> hshift[i], block_rect.y0() >> vshift[i],
              block_rect.xsize() >> hshift[i], block_rect.ysize() >> vshift[i]);
+    const auto& dc = dec_state->shared->dc->Plane(i);
+    JXL_CHECK(r[i].IsInside({0, 0, dc.xsize(), dc.ysize()}));
```
[the remark this answers is not in the document]
Comment (veluca93) [reply lib/jxl/dec_group.cc:263]: (also the variable should probably be behind a `#if JXL_ENABLE_ASSERT`)

File: lib/jxl/passes_state.cc [hunk 2]
```diff
@@ -42,10 +42,12 @@ Status InitializePassesSharedState(const FrameHeader& frame_header,
...
   shared->quant_dc = ImageB(frame_dim.xsize_blocks, frame_dim.ysize_blocks);
-  if (!(frame_header.flags & FrameHeader::kUseDcFrame) || encoder) {
-    shared->dc_storage =
-        Image3F(frame_dim.xsize_blocks, frame_dim.ysize_blocks);
-  } else {
+
+  shared->dc_storage = Image3F(frame_dim.xsize_blocks, frame_dim.ysize_blocks);
+  shared->dc = &shared->dc_storage;
+
+  bool use_dc = !!(frame_header.flags & FrameHeader::kUseDcFrame);
```
Comment (veluca93) [inline lib/jxl/passes_state.cc:49]: maybe `use_dc_frame`?

Comment (veluca93) [reply lib/jxl/passes_state.cc:49]: (which reminds me: shouldn't we allocate dc_storage only if not using a DC frame?)

File: lib/jxl/dec_group.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (eustas) [reply lib/jxl/dec_group.cc]: Done

File: lib/jxl/passes_state.cc (hunk 2 above)
Comment (eustas) [reply lib/jxl/passes_state.cc]: Done.

(Not unlikely)

Comment (veluca93) [reply lib/jxl/passes_state.cc]: Can we restore that logic and see if the buffer overflow is still resolved?

Comment (eustas) [reply lib/jxl/passes_state.cc]: Seems to work.
Also reallocated `dc_storage` to 0x0 image, so that problem would manifest itself earlier next time.
PTAL

**Review by veluca93 — APPROVED** [review]

**Pull request merged**